Jdb the great expansion

jdblair opened the pull request:
This changes the format of the OSC message so we can send 1 message and update a whole set of beads at once. This appears to fix our performance problems caused by trying to update 480 beads (the individual LEDs on the cross) b/c now they can be sent as 1 update.

Beads now have a "class" value that defines the name of the set. Currently there is rosary, cross and bases. The new format is /bead siib <class> <base> <count> <binary blob>. The base is the starting bead number. The count is the number of beads. The binary blob has 4 bytes per bead in the form r,g,b,brightness.

The pull request also includes the changes to oscled (the C++ controller) and the python simulator to correctly receive the new message format. oscled also includes changes to the configuration file format so we can define the type, or "interface class" of each interface, so we can track the different types of bead updates.

Commit:
color.set() now takes the alpha channel value into effect
we need to change the default alpha channel value to 1 for all the places where we pick colors to get the same behavior as before.

Commit:
Merge branch 'master' into jdb-alpha-channel to get all those sweet changes we just merged to master

Commit:
Merge branch 'jdb-vibration' into jdb-alpha-channel

Commit:
alpha channel experimentation

Commit:
added "xform" option to define a linear transformation of the RGB values, for adjusting levels and implementing low-power mode
the default xform option is not working correct

Commit:
implemented /led/xform method

Commit:
checking in b/c i'm going to conduct major surgery to see if i can speed things up

Commit:
I think I've taken care of the performance problem by changing the OSC message into a binary blob

Commit:
OMG, it fucking works.

this implements a new binary blob OSC message
the format is (base, length, blob), where:
base: base bead number (first bead in the byte array)
length: number of beads in the byte array
blob: a binary blob where each bead is the python struct format "!HHH", or 3 network byte order shorts (2 byte integers)
I chose 2 byte integers even though our LEDs are 1 byte devices b/c want to preserve more dynamic range in case we have time to get fancy with the 8 LEDs in each rosary bead

further, it splits all the LEDs in our system into 3 update groups, defined by the Updater object:
rosary beads: the original 60 beads
bases: the 9 base lights
cross: the 480 individual LEDs in the cross

Commit:
implement /bead/rosary method, receiving a blob type

Commit:
remove me some printfs, add some commentary

Commit:
just checkin' in

Commit:
finish new "interface class" option
this lets us label LED interfaces to separate sets of beads, each set with its own numbering
right now it only works with the /bead method and the blob payload
the format is:

/bead siib <class> <base> <length> <binary blob>

The binary blob is an array of 4 16-bit values: r, g, b, brightness

Commit:
remove debug printf()

jdblair commented on python/mp/color.py:
```diff
@@ -6,6 +6,7 @@ def __init__(self, r=0, g=0, b=0, a=0, name='color'):
         self.g = g  # green
         self.b = b  # blue
         self.a = a  # alpha channel (not yet used)
+        self.brightness = 0xff;  # brightness used by APA102 LED module
```
We're not using this yet, but the APA102 LED module has a 4th value that can be used to control the overall brightness level by controlling the current flowing to all three LEDs

jdblair commented on python/mp/rosary.py:
```diff
@@ -18,7 +19,7 @@
     """Bead represents a single rosary bead."""
     def __init__(self, index=0):
-        self.index = index
+        self.index = int(index)
```
this was an early experiment to speed things up, b/c I read that python3 iterators are slower if you don't force values to int().

jdblair commented on python/mp/rosary.py:
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
+
+        self.osc_client.send(msg.build())
+
+
```
This object contains the context needed to send an OSC message for a set of beads. This context is name (class), the bead_list for the set of beads, and the osc_client object.

jdblair commented on python/mp/rosary.py:
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
     def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
         self.beads = []
+        self.bases = []
+        self.cross = []
```
This is the 2 new sets of beads: bases and cross.

jdblair commented on python/mp/rosary.py:
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
         self.BEAD_COUNT=60
+        self.BASE_COUNT=9
+        self.CROSS_LED_COUNT=480
         self.run_mainloop = False
```
size of two new set of beads

jdblair commented on python/mp/rosary.py:
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
         # Available knobs to turn
         self.knobs = {}
+        self.updater_list = []
 
```
this holds the list of Updater objects that we step through in our mainloop

jdblair commented on python/mp/rosary.py:
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
+                                         bead_list=self.cross,
+                                         osc_client=self.osc_client))
+
 
```
create the three sets of beads and the associated Updater

jdblair commented on python/mp/rosary.py:
```diff
@@ -82,7 +132,9 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
+            'odd_ring': frozenset(self.beads[5:60:2]),
+            'base': frozenset(self.bases[0:self.BASE_COUNT]),
+            'cross': frozenset(self.cross[0:self.CROSS_LED_COUNT])
         }
```
two new set names: base and cross. we're going to want more, but this gets us started.

jdblair commented on python/mp/rosary.py:
```diff
@@ -122,8 +172,8 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
-        for bead in self.beads:
+    def beads_set_bgcolor(self, beads):
+        for bead in beads:
             bead.color.set(self.bgcolor)
```
reset all the beads in all the bead classes

jdblair commented on python/mp/rosary.py:
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
+                                       int(bead.color.g * 0xffff),
+                                       int(bead.color.b * 0xffff),
+                                       int(bead.color.brightness)
+            ))
```
This is where we pack the binary blob format: r,g,b,brightness in 4 binary bytes. The byte order is big endian (network byte order) b/c that's the tradition, even though it means byte-swapping on both ends of the wire.

GANDAMU replied:
If I understand correctly, we used to send /beadf with a bead ID and a color, and now we're sending /bead with a bead ID start, how many beads from there, and the color?

jdblair replied:
We're sending the starting bead, the number of beads and then a binary blob that contains the colors for all of those beads. It means we send 3 updates for all the "beads": 1 for the original rosary beads, 1 for the lights in the 9 base structures, and 1 for the 480 LEDs in the cross. It was too expensive to send 60 + 9 + 480 individual updates.

jdblair commented on python/pympsim/simulator.py:
```diff
@@ -41,15 +42,28 @@ def __init__(self, master):
+    #         print(r,g,b)
+    #         tk_rgb = "#%02x%02x%02x" % (r, g, b)
+    #         self.canvas.itemconfig(self.beads[bead], fill=tk_rgb)
+
```
I left this in here in case we need it again, even though its cruft.

jdblair commented on python/pympsim/simulator.py:
```diff
@@ -41,15 +42,28 @@ def __init__(self, master):
+                self.canvas.itemconfig(self.beads[num], fill=tk_rgb)
+                num += 1
+
 
```
Decode the binary blob

jdblair commented on python/pympsim/OSCserver.py:
```diff
@@ -19,6 +19,14 @@ def print_update(addr, args):
+
+    beadblob.append([base, length, blob])
+    queue.put(beadblob)
     beadblob = []
```
push the binary blob on the queue

jdblair commented on src/opi/PlatformSerialFactory.cpp:
```diff
@@ -8,13 +8,34 @@
+        ser.reset(new PlatformSerial(8, 10));
+        break;
+    }
 
```
These are the GPIO locations in the new interface board.
See https://oshpark.com/orders/9GaWwwGm

jdblair commented on src/oscled.json:
```diff
@@ -14,15 +14,24 @@
+                    "led_base": 0,
+                    "led_count": 8,
+                    "reversed": false
                 }
```
this is an example "base" class. all controllers will have at least one base attached on the third port.

jdblair commented on src/OSCServer.h:
```diff
@@ -35,6 +35,10 @@ class OSCServer {
+    static const int BEAD_BASES;
+    static const int BEAD_CROSS;
+
     int bind(std::shared_ptr<IPlatformSerial> const ser, OSCLedConfig::interface_config const &cfg);
```
whoops, i didn't end up using these.

Commit:
resolved conflicts when merging jdb-the-great-expansion

Commit:
whoops - checked in part of the conflict marker

Commit:
define all stations and interfaces

Commit:
now actually define all the interfaces

Commit:
Merge branch 'jdb-nested-effect-list' into jdb-alpha-channel

Conflicts:
	python/mp/effects/__init__.py
	python/mp/rosary.py

Commit:
work out some issues with alpha channel
also make some changes to casino so we can make it go faster

GANDAMU commented on python/mp/rosary.py:
```diff
@@ -231,28 +281,38 @@ def register_written_triggers(self):
+            
+        #bundle = bundle.build()
+        #self.osc_client.send(bundle)
+        self.osc_client.send(msg.build())
```
If I understand the point of this PR correctly, it's now the job of the Updater classes to call update() and set the colors on the individual beads? What do the changes here do then? Or rather, what does rosary.update() do now?

jdblair replied:
rosary.update() is cruft now. I should have removed it. Or we can keep it and I can move the code that calls updater there.

GANDAMU commented on python/mp/rosary.py:
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
+        self.updater_list.append(Updater(name='cross',
+                                         bead_list=self.cross,
+                                         osc_client=self.osc_client))
+
```
Shouldn't the bead class be taking a name arg too now? Unless I'm missing something, just reading this it seems self.bases[0], self.beads[0], and self.cross[0] would all refer to the same bead?

jdblair replied:
No, Bead should not take a name. self.bases[0], self.beads[0] and self.cross[0] do not refer to the same bead. There are 3 python lists of beads to help us keep them straight, but the effects never see these arrays. The effects get sets of beads, and we can pull those sets from any of the three sets of beads.

Right now if you mix beads from multiple lists in the same set we have a problem b/c they won't sort correctly. I'm "sorting" that out still. Hah. In reality I don't think we will want to combine, say, the cross and the rosary in the same effect so I'm not too concerned.

jdblair replied:
To put it another way, here are some examples:

```
# run 3phase_sine_wave on the cross
oscsend localhost 5006 /rosary/add_effect/name/3phase_sine_wave/bead_set/cross
# run throb on the bases
oscsend localhost 5006 /rosary/add_effect/name/throb/bead_set/bases
# run bounce on the rosary
oscsend localhost 5006 /rosary/add_effect/name/bounce/bead_set/rosary
```

GANDAMU reviewed (changes requested):
It all LOOKs good, for the most part, save for my point about the beads. But I can't manage to add any effects.

I'm hoping my one point had something to do with it though.

Commit:
added wheel and the start of roulette
cleaned up some stuff in the other effects
moved expose_effect_knobs() out of the mainloop, just call it when adding an effect

Commit:
Merge branch 'jdb-alpha-channel' into jdb-the-great-expansion

Conflicts:
	python/mp/effects/set_color.py
	python/mp/rosary.py

Commit:
arggg. left conflict markers in last commit.

Commit:
fix python simulator to receive the revised message format correctly. difference is:
8 bytes per bead, not 6
encode iface_class in a string arg, not in the method name

Commit:
remove debug print messages

Commit:
Merge branch 'jdb-linear-trans' into jdb-the-great-expansion

Conflicts:
	src/oscled.json

Commit:
add 'rosary' set
